control and planning scripts/planner.py

```diff
@@ -0,0 +1,204 @@
+			waypoints_out.poses.append(pose)
+		pub.publish(waypoints_out)
+	return
+	#once a waypoint is slided then new path is generated between itself and next global wp so we need to check 
+	#for collision on that path by arbitration and again checking new path if its feasible. 
+	#so the checking process dhould terminate and then restart with new pathafter sliding.
+	
+def slider( i, x_contour, y_contour, check_bool, xs, ys ):
+	global lookup
+	global slider_list
+	global waypoint_list_cropped
+	if slider_list[i]:
+		if i > 1 and i < len(waypoint_list_cropped)-2:
+			b1 = slider(i-1, x_contour, y_contour, 0, xs, ys )
+			if b1: 
+				b2 = slider(i+1, x_contour, y_contour, 0,xs, ys)
+				if b2:
+					slider_list[i] = False
+					b3 = slider(i, x_contour, y_contour, 1, xs, ys )
+			if not b1 & b2 & b3:
+				return False
+		else:
+			print('path not possible better stop here')
+			return False
+	if not slider_list[i]:
+		if i > 0 and i < len(waypoint_list_cropped)-1:
+			x_path = waypoint_list_cropped[i].x
+			y_path = waypoint_list_cropped[i].y
+			x_pre = waypoint_list_cropped[i-1].x
+			y_pre = waypoint_list_cropped[i-1].y
+			slope = math.atan2((y_path - y_pre),(x_path - x_pre))
+			x_pos = x_path + d* math.sin(slope)
+			y_pos = y_path + d* math.cos(slope)
+			x_neg = x_path - d* math.sin(slope)
+			y_neg = y_path - d* math.cos(slope)
+			a, b, c = line_eq(x_path, y_path, x_pre, y_pre)
+			s1 = (a*x_pos + b*y_pos - c)>0			# basically if 2 points lie on same sid eof line then they give same sign when substituted in line eq
+			s2 = (a*x_contour + b*y_contour - c)>0
+			direction = (s1 & s2) + ((not s1) & (not s2)) #xor gate
+			if not direction :                    # means 2 points lie on opp. dir of path
+				waypoint_list_cropped[i].x , waypoint_list_cropped[i].y = x_pos , y_pos
+			else:
+				waypoint_list_cropped[i].x , waypoint_list_cropped[i].y = x_neg , y_neg
+			slider_list[i] = True
+			if check_bool:
+				collision_checker(xs, ys)
+		else:
+			print('too less points and that too in collision u better stop')
+			return False
+	return True
+
+def line_eq(x1, y1, x2, y2):
+	a = y2 - y1
+	b = x1 - x2
+	c = a*x1 + b*y1
+	print('returning line eq')
+	return a, b, c
+
+def get_nearest_global_point(waypoint_list, xc, yc):
+	min_dist_sq = 1000
+	for i in range(len(waypoint_list)-1):
+		x_path = waypoint_list[i].x
+		y_path = waypoint_list[i].y
+		dist_sq = (xc - x_path)**2 + (yc - y_path)**2
+		if dist_sq < min_dist_sq:
+			near_index = i
+	print('returning nearest index ',near_index)
+	return near_index
+
+# def path_optimiser(new_waypoint):
+
+# 	pub.publish(waypoints_out)
+
+
+
+############################################ section 4 callbacks	
+def callback_odom(data_pose):
+	global lookup
+	global slider_list
+	global waypoint_list_cropped
+	curr_pose = PoseStamped()
+	#print(len(lookup))
+	curr_pose = data_pose
+	waypoint_list = []
+	data = rospy.wait_for_message("/global_path", Path)
+	for i in range(len(data.poses)):
```

Comment:
it can be done in one line like `waypoint_list.append(data.poses[i].pose.position for i in range(len(data.poses))`

Reply:
seems like it should work

Reply:
Does that make code more efficient? or just changing the syntax?

Reply:
improves the runtime marginally. You can search for 'list comprehensions vs for loops'

Reply:
Done!!